src/librustc_resolve/late.rs

```diff
@@ -544,6 +544,50 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
             self.visit_where_predicate(p);
         }
     }
+
+    fn visit_generic_arg(&mut self, arg: &'tcx GenericArg) {
+        debug!("visit_generic_arg({:?})", arg);
+        match arg {
+            GenericArg::Type(ref ty) => {
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    // A path type argument could be an incorrectly parsed generic const argument.
+                    // We double check just to be sure by trying to resolve in the type namespace
+                    // first, and if that fails we try again in the value namespace. If resolution
+                    // in the value namespace succeeds, we have an improperly parsed generic const
+                    // argument on our hands.
+                    if path.segments.len() == 1 {
+                        if let None = self.resolve_qpath(
```

Comment:
Speculative resolution here should use `resolve_ident_in_lexical_scope` instead of `resolve_qpath`.
See how `PatKind::Ident` are disambiguated using speculative resolution in `resolve_pattern_inner` for an example.
(`resolve_qpath` may leave side effects.)

Reply:
I actually first started with using `resolve_ident_in_lexical_scope`, and swapped over to `resolve_qpath` rather late into these changes. I can fairly easily swap back.